File: app/src/main/java/alexis/tvrs/rickandmortyuniverse/ui/characters/AsyncDownloadCharacter.java
package alexis.tvrs.rickandmortyuniverse.ui.characters;

import android.os.AsyncTask;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

public class AsyncDownloadCharacter extends AsyncTask<Object,Void, Character> {
    URL url;
    ArrayList<Character> list;
    CharacterGridAdapter characterGridAdapter;

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
    }

    @Override
    protected Character doInBackground(Object[] objects) {
        Character character = null;
        int characterNb = (int) objects[0];
        list = (ArrayList<Character>) objects[1];
        characterGridAdapter = (CharacterGridAdapter) objects[2];
        String jsonURL = "https://rickandmortyapi.com/api/character/"+characterNb;
        try {
            url = new URL(jsonURL);
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            if (urlConnection.getResponseCode() == HttpURLConnection.HTTP_OK) { InputStreamReader isr = new InputStreamReader(urlConnection.getInputStream());
                BufferedReader input = new BufferedReader(isr);
                String jsonStr = input.readLine();
                JSONObject jsonObject = new JSONObject(jsonStr);

                int id = jsonObject.getInt("id");
                String image = jsonObject.getString("image");
                String name = jsonObject.getString("name");
                String species = jsonObject.getString("species");
                String status = jsonObject.getString("status");
                String gender = jsonObject.getString("gender");
                JSONObject originObject = jsonObject.getJSONObject("origin");
                String origin = originObject.getString("name");
                JSONObject lastLocationObject = jsonObject.getJSONObject("location");
                String lastLocation = lastLocationObject.getString("name");

                character = new Character(id,image,name,status,species,gender,origin,lastLocation);
                character.SetLastLocationUri(lastLocationObject.getString("url"));
                character.SetOriginUri(originObject.getString("url"));
                input.close();
            }
        } catch (IOException | JSONException e) {
            e.printStackTrace();
        }
        return character;
    }

    @Override
    protected void onPostExecute(Character characterRickMorty) {
        super.onPostExecute(characterRickMorty);
        list.add(characterRickMorty);
        characterGridAdapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/alexis/tvrs/rickandmortyuniverse/ui/characters/AsyncDownloadCharacter.kt
package alexis.tvrs.rickandmortyuniverse.ui.characters

import android.os.AsyncTask
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.*

class AsyncDownloadCharacter : AsyncTask<Any, Void?, Character?>() {
    var url: URL? = null
    var list: ArrayList<Character?>? = null
    var characterGridAdapter: CharacterGridAdapter? = null
    override fun onPreExecute() {
        super.onPreExecute()
    }

    override fun doInBackground(objects: Array<Any>): Character? {
        var character: Character? = null
        val characterNb = objects[0] as Int
        list = objects[1] as ArrayList<Character?>
        characterGridAdapter = objects[2] as CharacterGridAdapter
        val jsonURL = "https://rickandmortyapi.com/api/character/$characterNb"
        try {
            url = URL(jsonURL)
            val urlConnection = url!!.openConnection() as HttpURLConnection
            if (urlConnection.responseCode == HttpURLConnection.HTTP_OK) {
                val isr = InputStreamReader(urlConnection.inputStream)
                val input = BufferedReader(isr)
                val jsonStr = input.readLine()
                val jsonObject = JSONObject(jsonStr)
                val id = jsonObject.getInt("id")
                val image = jsonObject.getString("image")
                val name = jsonObject.getString("name")
                val species = jsonObject.getString("species")
                val status = jsonObject.getString("status")
                val gender = jsonObject.getString("gender")
                val originObject = jsonObject.getJSONObject("origin")
                val origin = originObject.getString("name")
                val lastLocationObject = jsonObject.getJSONObject("location")
                val lastLocation = lastLocationObject.getString("name")
                character = Character(id, image, name, status, species, gender, origin, lastLocation)
                character.SetLastLocationUri(lastLocationObject.getString("url"))
                character.SetOriginUri(originObject.getString("url"))
                input.close()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return character
    }

    override fun onPostExecute(characterRickMorty: Character?) {
        super.onPostExecute(characterRickMorty)
        list!!.add(characterRickMorty)
        characterGridAdapter!!.notifyDataSetChanged()
    }
}